src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/TelemetryMiddleware.cs

```diff
@@ -0,0 +1,33 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using Microsoft.AspNetCore.Http;
+
+namespace OpenTelemetry.Instrumentation.AspNetCore.Implementation;
+
+/// <summary>
+/// Telemetry middleware.
+/// </summary>
+/// <remarks>
+/// Initializes a new instance of the <see cref="TelemetryMiddleware"/> class.
+/// </remarks>
+/// <param name="options">trace options.</param>
+public class TelemetryMiddleware(AspNetCoreTraceInstrumentationOptions options) : IMiddleware
+{
+    internal readonly HttpInListener HttpInListener = new HttpInListener(options);
+
+    /// <inheritdoc/>
+    public async Task InvokeAsync(HttpContext context, RequestDelegate next)
+    {
+        try
+        {
+            this.HttpInListener.OnEventWritten("Microsoft.AspNetCore.Hosting.HttpRequestIn.Start", context);
```

Comment:
if using middleware approach, then HttpInListener should not be used, as that'll incur the cost of DS callbacks, which a middleware can avoid

Reply:
it won't, It does not enable DiagnosticSourceSubscriber which we do here https://github.com/open-telemetry/opentelemetry-dotnet/blob/a73657476406c6ae1bdd6fd30b23e869646ada91/src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs#L29